### PR TITLE
stop moving window after N steps

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Sometimes for bunch-bunch or bunch-target collision it could be useful to allow to stop the
moving windows. This option is now available via `stopWindow`.

- add cli option `stopWindow`
- refactor `MovingWindow`: add method `setEndSlideOnStep` and `isEnabled`

# Tests

- [x] LWFA with `-m`
- [x] LWFA `-m --stopWindow 2000`
- [x] LWFA no sliding window

### CODE REVIEW

**Comment by PrometheusPi** [comment]
@psychocoderHPC why is the `-w` flag needed when stopping while the `-m` is needed when not stopping- Shouldn't it be the same flag to activate a moving window?

File: include/picongpu/simulationControl/MovingWindow.hpp [hunk 1]
```diff
@@ -181,34 +186,55 @@ class MovingWindow
     void incrementSlideCounter(const uint32_t currentStep)
     {
         // do not slide twice in one simulation step
-        if (slidingWindowActive==true && lastSlideStep < currentStep)
+        if (currentStep < endSlidingOnStep && lastSlideStep != currentStep)
```
Comment (ax3l) [inline include/picongpu/simulationControl/MovingWindow.hpp]: you just reverted #2774 with "!="

**Comment by ax3l** [comment]
@psychocoderHPC in case you are at work, I reviewed this PR :)

File: include/picongpu/simulationControl/MovingWindow.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/picongpu/simulationControl/MovingWindow.hpp]: well spotted! thx

**Comment by psychocoderHPC** [comment]
> docs/TBG_macros.cfg needs an update as well

@ax3l  the doc update is included
https://github.com/ComputationalRadiationPhysics/picongpu/pull/2792/files#diff-6ed48ed85428745c2f464a9f3643f879R94

**Comment by psychocoderHPC** [comment]
> @psychocoderHPC why is the -w flag needed when stopping while the -m is needed when not stopping- Shouldn't it be the same flag to activate a moving window?

@PrometheusPi  I fixed the initial description. `-w` is not needed it must be a `-m` I pressed the wrong key on my keyboard (they are so near by each other ;-)

File: include/picongpu/simulationControl/MovingWindow.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/picongpu/simulationControl/MovingWindow.hpp]: fixed

File: include/picongpu/fields/FieldManipulator.hpp [hunk 2]
```diff
@@ -83,7 +83,7 @@ class FieldManipulator
                 }
 
                 /* if sliding window is active we disable absorber on bottom side*/
-                if (MovingWindow::getInstance().isSlidingWindowActive() && i == BOTTOM) continue;
+                if (MovingWindow::getInstance().isSlidingWindowActive(currentStep) && i == BOTTOM) continue;
```
Comment (ax3l) [inline include/picongpu/fields/FieldManipulator.hpp]: @psychocoderHPC why did you change this as well?

Comment (ax3l) [reply include/picongpu/fields/FieldManipulator.hpp]: forget about it, I saw it below... makes sense

File: include/picongpu/simulationControl/MovingWindow.hpp [hunk 3]
```diff
@@ -181,34 +186,55 @@ class MovingWindow
     void incrementSlideCounter(const uint32_t currentStep)
     {
         // do not slide twice in one simulation step
-        if (slidingWindowActive==true && lastSlideStep < currentStep)
+        if (currentStep < endSlidingOnStep && lastSlideStep < currentStep)
```
Comment (ax3l) [inline include/picongpu/simulationControl/MovingWindow.hpp]: since `endSlidingOnStep` defaults to `0u` and `currentStep` is `>=0` this is fine. But by the logic that was here before, you should first check if `isEnabled()` is true.

File: include/picongpu/fields/FieldManipulator.hpp (hunk 2 above)
Comment (psychocoderHPC) [reply include/picongpu/fields/FieldManipulator.hpp]: Because without the time step it is not possible to check if the moving window is stopped.
As soon as the window stops the bottom absorber is enabled else the laser get reflected

File: include/picongpu/simulationControl/MySimulation.hpp [hunk 4]
```diff
@@ -117,6 +117,7 @@ class MySimulation : public SimulationHelper<simDim>
     cellDescription(nullptr),
     initialiserController(nullptr),
     slidingWindow(false),
+    endSlidingOnStep(0),
```
Comment (ax3l) [inline include/picongpu/simulationControl/MySimulation.hpp]: this should be `-1` for consistency with the default value. Or you make it `uint32_t` as the usage in `MovingWindow` and put `0` as a default, which I find reasonable a well as a special value (do *not* stop at zero).

File: include/picongpu/simulationControl/MovingWindow.hpp (hunk 3 above)
Comment (psychocoderHPC) [reply include/picongpu/simulationControl/MovingWindow.hpp]: Sounds good than it is more cleare

File: include/picongpu/fields/FieldManipulator.hpp (hunk 2 above)
Comment (ax3l) [reply include/picongpu/fields/FieldManipulator.hpp]: Yes, overlooked this

File: include/picongpu/simulationControl/MySimulation.hpp (hunk 4 above)
Comment (psychocoderHPC) [reply include/picongpu/simulationControl/MySimulation.hpp]: To enable the moving window we have a bool switch. Therefore I set the default to `disbale/0`
`uint32_t` is not useful because than I need to show the user a magic large number as default in the help (https://github.com/ComputationalRadiationPhysics/picongpu/pull/2792/files). It is more common to have `-1` as unlimited.

Comment (ax3l) [reply include/picongpu/simulationControl/MySimulation.hpp]: But in `MovingWindow` you set it to `0`. I would be fine to make it all-`0` as default as you got it there.

Comment (psychocoderHPC) [reply include/picongpu/simulationControl/MySimulation.hpp]: The object `MovingWindow` is Independent of `MySimulation` there is no need that internal variables of two different objects fits in his default value.
IMO `-1` fits much better to `--stopWindow step` instead `zero`

Comment (ax3l) [reply include/picongpu/simulationControl/MySimulation.hpp]: Sure, but following that logic, shouldn't `MovingWindow` also use `-1` for "do not move"? ;-)

Comment (ax3l) [reply include/picongpu/simulationControl/MySimulation.hpp]: Also, this has a bug. If a user *would* truly request `--stopWindow 0` then it would still not stop at zero. Therefore, make the cmd line interface with `0 means not stopping` as well for simplicity?

Comment (ax3l) [reply include/picongpu/simulationControl/MySimulation.hpp]: update, no wait, it works as intended.

File: docs/TBG_macros.cfg [hunk 5]
```diff
@@ -90,6 +90,8 @@ TBG_periodic="--periodic 1 0 1"
 # Enables moving window (sliding) in your simulation
 TBG_movingWindow="-m"
 
+# stop the moving window after given simulation step
+TBG_stopWindow="--stopWindow 1337"
 ################################################################################
```
Comment (ax3l) [inline docs/TBG_macros.cfg]: pls give me an empty line :)

Comment (psychocoderHPC) [reply docs/TBG_macros.cfg]: You and your white space fetish^^ ( I will add the empty line)

Comment (ax3l) [reply docs/TBG_macros.cfg]: The - symmetry - is - broken! xD

Comment (psychocoderHPC) [reply docs/TBG_macros.cfg]: I forgot to push my changes
